Client/Frontend/Browser/HomePanel/NewTabPageBackgroundDataSource.swift

```diff
@@ -0,0 +1,178 @@
+    // This can 'easily' be adjusted to support both sets by switching to String, and using filePath to identify uniqueness.
+    private var lastBackgroundChoices = [Int]()
+    
+    func newBackground() -> (Background, Sponsor?)? {
+        if !Preferences.NewTabPage.backgroundImages.value { return nil }
+        
+        // Identifying the background array to use
+        let (backgroundSet, useSponsor) = { () -> ([NewTabPageBackgroundDataSource.Background], Bool) in
+            // Determine what type of background to display
+            let attemptSponsored = Preferences.NewTabPage.backgroundSponsoredImages.value
+                && backgroundRotationCounter == NewTabPageBackgroundDataSource.sponsorshipShowValue
+                && !PrivateBrowsingManager.shared.isPrivateBrowsing
+            
+            if let sponsoredWallpapers = sponsor?.wallpapers, attemptSponsored {
+                sponsoredBackgroundRotationIndex += 1
+                // Force a max, and wrap back down if it is hit.
+                sponsoredBackgroundRotationIndex %= sponsoredWallpapers.count
+                return (sponsoredWallpapers, true)
+            }
+            return (standardBackgrounds, false)
+        }()
+        
+        if backgroundSet.isEmpty { return nil }
+        
+        // Choosing the actual index / item to use
+        let backgroundIndex = { () -> Int in
+            if useSponsor {
+                return self.sponsoredBackgroundRotationIndex
+            }
+            
+            let availableRange = 0..<backgroundSet.count
+            // This takes all indeces and filters out ones that were shown recently
+            let availableBackgroundIndeces = availableRange.filter {
+                !self.lastBackgroundChoices.contains($0)
+            }
+            // Chooses a new random index to use from the available indeces
+            // -1 will result in a `nil` return
+            let chosenIndex = availableBackgroundIndeces.randomElement() ?? -1
+            assert(chosenIndex >= 0, "NTP index was nil, this is terrible.")
+            assert(chosenIndex < backgroundSet.count, "NTP index is too large, BAD!")
+
+            // This index is now added to 'past' tracking list to prevent duplicates
+            self.lastBackgroundChoices.append(chosenIndex)
+            // Trimming to fixed length to release older backgrounds
+            self.lastBackgroundChoices = self.lastBackgroundChoices.suffix(NewTabPageBackgroundDataSource.numberOfDuplicateAvoidance)
+            return chosenIndex
+        }()
+        
+        // Force back to `0` if at end
+        backgroundRotationCounter %= NewTabPageBackgroundDataSource.sponsorshipShowRate
+        // Increment regardless, this is a counter, not an index, so smallest should be `1`
+        backgroundRotationCounter += 1
+        
+        // Item is returned based on our special index.
+        
+        return (backgroundSet[backgroundIndex], useSponsor ? sponsor : nil)
```

Comment:
Can use `backgroundSet[safe: backgroundIndex]` so it never crashes when out of bounds

Reply:
That returns an optional though. There should be no reason for this to crash, as it is chosen from the range, but we could build in more safeguards?

Reply:
fine by me